platform/pulp/tasking/system.py

```diff
@@ -0,0 +1,474 @@
+                              one task reserving this same string can happen at a time.
+        :type  resource_id:   basestring
+
+        :param tags:          A list of tags (strings) to place onto the task, used for searching
+                              for tasks by tag
+        :type  tags:          list
+
+        :param group_id:      The id to identify which group of tasks a task belongs to
+        :type  group_id:      uuid.UUID
+
+        :return:              An AsyncResult instance as returned by Celery's apply_async
+        :rtype:               celery.result.AsyncResult
+        """
+        # Form a resource_id for reservation by combining given resource type and id. This way,
+        # two different resources having the same id will not block each other.
+        resource_id = ":".join((resource_type, resource_id))
+        inner_task_id = str(uuid.uuid4())
+        task_name = self.name
+        tag_list = kwargs.get('tags', [])
+        group_id = kwargs.get('group_id', None)
+
+        # Set the parent attribute if being dispatched inside of a Task
+        parent_arg = self._get_parent_arg
+
+        # Create a new task status with the task id and tags.
+        task_status = TaskStatus(id=inner_task_id, state=TaskStatus.WAITING, tags=tag_list,
+                                 group=group_id, **parent_arg)
+        task_status.save()
+        _queue_reserved_task.apply_async(args=[task_name, inner_task_id, resource_id, args, kwargs],
+                                         queue=RESOURCE_MANAGER_QUEUE)
+        return AsyncResult(inner_task_id)
+
+    def apply_async(self, *args, **kwargs):
+        """
+        A wrapper around the super() apply_async method. It allows us to accept a few more
+        arguments than Celery does for our own purposes, listed below. It also allows us
+        to create and update task status which can be used to track status of this task
+        during it's lifetime.
+
+        :param queue:       The queue that the task has been placed into (optional, defaults to
+                            the general Celery queue named 'celery'.)
+        :type  queue:       basestring
+
+        :param tags:        A list of tags (strings) to place onto the task, used for searching for
+                            tasks by tag
+        :type  tags:        list
+
+        :param group_id:    The id that identifies which group of tasks a task belongs to
+        :type group_id:     uuid.UUID
+
+        :return:            An AsyncResult instance as returned by Celery's apply_async
+        :rtype:             celery.result.AsyncResult
+        """
+        routing_key = kwargs.get('routing_key',
+                                 defaults.NAMESPACES['CELERY']['DEFAULT_ROUTING_KEY'].default)
+        tag_list = kwargs.pop('tags', [])
+        group_id = kwargs.pop('group_id', None)
+        async_result = super(UserFacingPulpTask, self).apply_async(*args, **kwargs)
```

Comment:
It'd be super cool if there was a check here for the Django setting `profile = True` or something and then start a cProfile that is stored in MEDIA_ROOT/profiles/<uuid> or something. That way users can easily generate profile information.

I know that's not really the point of this PR, but it'd be cool.